### PR TITLE
Use official golang images for windows

### LINKED ISSUE #41177: Cannot build from sources
**Description**

Currently contribution guide does not work if build on Windows. Currently there #39097 attempts to fix the problem, but it going nowhere. #39906 is some latest attempt which was closed. Seems to be currently there need to adhoc changes in the `Dockerfile.windows` to matching version and do that manually. If there no decision so far, maybe better update documentation to include steps necessary to build Docker locally?

**Steps to reproduce the issue:**
1. git clone
2. docker build -t nativebuildimage -f .\Dockerfile.windows .

**Describe the results you received:**
manifest for microsoft/windowsservercore:latest not found: manifest unknown: manifest unknown

**Describe the results you expected:**
Actually start building

**Additional information you deem important (e.g. issue happens only occasionally):**
- https://techcommunity.microsoft.com/t5/containers/windows-server-2019-now-available/ba-p/382430

### PR DESCRIPTION
**Pull request by thaJeztah**
Was curious if this would work 😅 
Will probably have to update the Makefile (if that's used on Windows)

fixes https://github.com/moby/moby/issues/41177

### CODE REVIEW

**Comment by olljanat** [comment]
> Will probably have to update the Makefile (if that's used on Windows)

Afaik, mkefile is not used on Windows.

Btw. Looks that some command line tools are now missing and that why build fails.

**Comment by thaJeztah** [comment]
> Btw. Looks that some command line tools are now missing and that why build fails.

Yes, I saw that failure; was trying to figure out where / how those were installed previously.

I don't have a Windows machine locally (hence pushing it as a PR 😇) and no real experience with PowerShell, so this was a bit of a "play around" thing yesterday evening to see if this could be simplified a bit (there's other parts in the script that I was looking at as well)..

I'll look a bit further in the weekend, or if I have a couple of minutes in between, but if you have suggestions / ideas, all welcome! 🤗 🎉

**Comment by thaJeztah** [comment]
> Afaik, mkefile is not used on Windows.

That's what I thought, but I got curious, because the `DOCKERFILE` env-var was set in the Makefile; https://github.com/moby/moby/blob/aea6fdf3d340835a1b0af208839ce42ace3a5b89/Makefile#L7-L9

Which made me think there are situations where it _could_ be used (e.g., in a Bash shell on Windows)

The DOCKERFILE env-var is set inside https://github.com/moby/moby/blob/aea6fdf3d340835a1b0af208839ce42ace3a5b89/hack/make/.detect-daemon-osarch#L37-L43

(where I noticed there's quite some stuff in that file that's no longer needed; I'll open a PR for removing some of the cruft)

**Comment by olljanat** [comment]
Actually. I think that lot of that stuff is most probably there because of RS1 (Windows Server 2016).
RS5 (Windows Server 2019) have support for Linux subsystem so most probably build process can be simplified quite a lot on that side.

**Comment by thaJeztah** [comment]
Note to self: need to add `golang` to the exclusion list for `Nuke-Everything`; see 
https://github.com/moby/moby/pull/38419#issuecomment-449599825

```
21:54:58 Untagged: golang:1.11.4-windowsservercore
```

**Comment by olljanat** [comment]
@thaJeztah I created PR https://github.com/thaJeztah/docker/pull/1 for you which adds busybox so at least some of these tests should works after you merge it.

I will also start testing https://docs.microsoft.com/en-us/windows/wsl/install-on-server to see if it helps us to minimize number of Windows specific hacks on CI.

**Comment by olljanat** [comment]
Hmm. It looks to be that https://www.nuget.org/api/v2/package/GitForWindows/2.11.1 actually contains another version binaries like *tar.exe* but I don't understand why it matters...

**Comment by thaJeztah** [comment]
I guess it's shelling out in these tests; https://github.com/moby/moby/blob/ea3ac621e3bf1c991c72f9f97defe0f4fd98be9b/pkg/archive/archive_test.go#L122-L135

File: Dockerfile.windows [hunk 1]
```diff
@@ -150,107 +150,56 @@
...
-  `
-  Write-Host INFO: Creating source directory...; `
-  New-Item -ItemType Directory -Path C:\go\src\github.com\docker\docker | Out-Null; `
-  `
   Write-Host INFO: Configuring git core.autocrlf...; `
   C:\git\cmd\git config --global core.autocrlf true; `
   `
-  Write-Host INFO: Completed
+  Write-Host 'INFO: Creating temp folder for unit tests...'; `
+  New-Item -ItemType Directory -Path C:\tmp; `
```
Comment (olljanat) [inline Dockerfile.windows]: It looks to be that tar.exe which run inside of busybox need to be able create archives to /tmp and those need to be available on user temp so replace this line with:
```
cmd /c mklink /D C:\tmp %TEMP%
```
and those archive tests will start working (or at least I got them working on my test machine).

**Comment by jhowardmsft** [comment]
@thaJeztah  Yeah, while you may be able to get this to work for official builds, it absolutely will break anyone running process isolation on nightly and probably insider builds (which is a huge number of people internally and swathes of test infrastructure).  The reason being that for process isolation, the kernel of the container image must match that of the host, and the official golang image isn't obviously built against a known matching base image.

You may be able to get away with copying the binaries and source directly out of the official image using `docker cp` and multi-stage, but if I'm reading this PR correctly (only a quick skim as first day back in office after almost a month, and brain isn't entirely engaged yet), attempting to run it as the base `FROM` image will definitely cause chaos internally 😨

**Comment by olljanat** [comment]
Process isolation should work with stable versions just fine as used golang:1.11.4-windowsservercore "image" looks to be actually multi version manifest (downloaded using [this script](https://github.com/TomasTomecek/download-manifest-from-dockerhub))
```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 3246,
      "digest": "sha256:7062dee2ed9264f76d1ae8a9b633638a1813dff82eeb0ed899218c165f0cc8f4",
      "platform": {
        "architecture": "amd64",
        "os": "windows",
        "os.version": "10.0.14393.2665"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 3245,
      "digest": "sha256:b281b734ab4edf68d2f0b230ad95330b294d613ba2672462949f534b137f2546",
      "platform": {
        "architecture": "amd64",
        "os": "windows",
        "os.version": "10.0.16299.846"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 3244,
      "digest": "sha256:cb9ddf6a702349bf79e9e6e6afcbb9b5f7a84d18dc38aa5de9b2b4addb5c2adc",
      "platform": {
        "architecture": "amd64",
        "os": "windows",
        "os.version": "10.0.17134.469"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "size": 3245,
      "digest": "sha256:0d1c1bc8bf89241d0a38a40d509fca97e966024c1026753021d67dd2ff55dbf9",
      "platform": {
        "architecture": "amd64",
        "os": "windows",
        "os.version": "10.0.17763.194"
      }
    }
  ]
}
```
but nighly/insider versions of course would need to be handled on some other way.

Anyway, I like of this idea as current build process is super complex and very hard to understand.

**Comment by jhowardmsft** [comment]
Right, but if you understood quite how many branches internally there are of Windows from the various teams, and realise Windows is built nightly, combined with the number of test suites which invoke the moby CI to make sure folks changing Windows aren't breaking containers, it's a very scary prospect just saying "would need to be handled some other way". It's critical that it continues to work for nightly and insiders - we absolutely can't have that broken.

**Comment by olljanat** [comment]
Hmm. It is same time good to hear that same build process is used with nightly and insider builds inside of Microsoft too but also bit sad that it also limits what can be done here :confused: 

> it's a very scary prospect just saying "would need to be handled some other way". It's critical that it continues to work for nightly and insiders - we absolutely can't have that broken.

On that case I think that only option is actually create new Dockerfile where we put this content when we get those tests working with it.

I'm especially interested to see if we can have Dockerfile which actually only create image which only contains build tools and scripts and then CI build would actually run even *git clone* inside of it before starting the build. Then it would much easier to create new build server(s) as only requirement would be to have Docker installed. It would also allow us to use some other CI tools instead of Jenkins (related to #38100)

> The reason being that for process isolation, the kernel of the container image must match that of the host

Btw. What is reason that it is not possible to make Windows working with kernel versions with process isolation like Linux does?

**Comment by olljanat** [comment]
> OK, but I think you are going to have an incredibly hard time supporting the current matrix across-the-board, and have it working from RS1 through RS5 (technically RS3 and RS4 are still supported), as well as the nightly 19H1 builds.

I was actually thinking about different kind of approach where we can keep existing Jenkins jobs as is but then enable new build which run only on RS5 as parallel of them but where we can easier enable those missing integration tests (I'm especially interested about Swarm tests as Windows swarm support looks to have many weaknesses which does not exist on Linux). That would also allow user to run those tests easier on development machine and github forks.

**Comment by jhowardmsft** [comment]
So you're suggesting building a parallel infrastructure? Ouch! The missing integration tests absolutely can be added - I had a proof of concept of that previously. It wasn't that much work. What's preventing you currently running on development machines/forks? I do that daily here on arbitrary builds, both released and nightly as part of my daily workflow. (That said, I haven't looked at Swarm at all, or it's requirements, but I suspect that may be even more problematic if you need CI automated across multiple machines)

**Comment by olljanat** [comment]
I'm not actually suggesting anything. Just thinking loudly about options.

Biggest issue what I can see on current mode if that it is too time consuming to setup new build machine which can be used with github forks. If you can share some script which is able to setup needed components example top of Azure "Windows Server 2019 Datacenter Server Core with Containers" template it would be very useful.

One option for Swarm tests would be btw enable some to Swarmkit side too. I shared some info how to do that on this PR https://github.com/docker/swarmkit/pull/2691 Especially on RS3+ it is easy as you can run swarmd.exe inside of container with npipe mount and have multiple copies of them running on single node.

**Comment by olljanat** [comment]
I guess that you need to update https://github.com/thaJeztah/docker/blob/f2013e0bd1170a55833ab7f7b065280d3e5cbba9/hack/ci/windows.ps1#L423 using value *Continue* not it does not tell why docker build fails.

Also I think that *--build-arg=GO_VERSION* is not needed as those values are hard coded on Dockerfile.

**Comment by thaJeztah** [comment]
> I guess that you need to update https://github.com/thaJeztah/docker/blob/f2013e0bd1170a55833ab7f7b065280d3e5cbba9/hack/ci/windows.ps1#L423 using value Continue not it does not tell why docker build fails.

I see the same pattern in other locations, and it looks like the actual error-handling is done manually in this line below that;

```powershell
if (-not($LastExitCode -eq 0)) {
```

Perhaps that was done to not make the whole script error out?


> Also I think that --build-arg=GO_VERSION is not needed as those values are hard coded on Dockerfile.

That allows optionally overriding the version by setting a `GO_VERSION` in the current environment (which can be useful, e.g. to test against other versions in CI, or when building packages without having to make changes upstream)

**Comment by olljanat** [comment]
> Perhaps that was done to not make the whole script error out?

Probably (and afaiu it was specially needed by Jenkins) but I was meaning that now build fails without printing any error and I have used Continue value earlier to get error details printed to log.

**Comment by thaJeztah** [comment]
This is green on RS5. Restarting CI with RS1 enabled (it's not enabled by default, and timed-out on previous runs, so let me try again)
